lib/SILOptimizer/Analysis/MemoryBehavior.cpp

```diff
@@ -240,27 +240,34 @@ MemBehavior MemoryBehaviorVisitor::visitApplyInst(ApplyInst *AI) {
 
   MemBehavior Behavior = MemBehavior::None;
 
-  // We can ignore mayTrap().
-  if (ApplyEffects.mayReadRC() ||
-      (InspectionMode == RetainObserveKind::ObserveRetains &&
-       ApplyEffects.mayAllocObjects())) {
-    Behavior = MemBehavior::MayHaveSideEffects;
+  if (all_of(enumerate(AI->getArgumentOperands()),
+             [&](llvm::detail::enumerator_impl<llvm::MutableArrayRef<swift::Operand> >::result_pair<swift::Operand &> op) -> bool {
+               return op.Value.get() == V && AI->getSubstCalleeType()->getParameters()[op.Index].isIndirectInGuaranteed();
+             })) {
+    Behavior = MemBehavior::MayRead;
   } else {
```

Comment:
nitpick: you can the 'else' with the following 'if' in one line to save one indentation level:
  } else if (...) {